### ISSUE TITLE
useClientRequest returns state for previous request

### ISSUE THREAD

**Issue by Joezo** [nearform/graphql-hooks#37]
This cropped up when testing out `useQuery` with different variables after the initial request. It looks like `useClientRequest` returns data the second time it's called, regardless of the query/options that were passed in.

**Comment by bmullan91** [comment]
Yep I discovered this recently too. In a normal component we could use `getDerivedStateFromProps` to ensure the state is correct. In hooks things are a little different, the `initialState` argument to `useReducer(reducer, initialState)`is only used for the first render.

Here's my proposed solution:

```js
function useClientRequest(query, initialOpts = {}) {
  // removed some code for clarity...

  const initalState = {
    ...(intialCacheHit ? intialCacheHit : client.generateResult({})),
    operation,
    cacheHit: !!intialCacheHit,
    loading: !intialCacheHit
  };

  const [state, dispatch] = React.useReducer(reducer, initalState);

  // compare operations
  // NOTE - it'd be nice to use the fnv1a key from memcache for comparison 
  if (JSON.stringify(state.operation) !== JSON.stringify(operation)) {
    dispatch({
      type: actionTypes.RESET_STATE,
      initalState
    });
  }
```

**Comment by bmullan91** [comment]
This might also be related: https://github.com/facebook/react/issues/14849

**Comment by gaearon** [comment]
React bug is fixed in 16.8.3.
